sdk/nodejs/runtime/resource.ts

```diff
@@ -63,7 +65,19 @@ export function readResource(res: Resource, t: string, name: string, props: Inpu
     log.debug(`Reading resource: id=${id}, t=${t}, name=${name}`);
 
     const monitor: any = getMonitor();
-    const resopAsync = prepareResource(label, res, true, props, opts);
+
+    // Get the ResourceResolverOperation for this resource.  Note: we grab the parent URN eagerly
+    // and we pass it in.  This is because resource calling into [readResource] will potentially
+    // update its URN immediately after calling us.  Because prepareResource performs async
+    // operations we may see the URN *post* update which is not what we want.  Indeed, this would
+    // cause cycles as urn is updated to to be dependent on children.  So we must grab the URN
+    // before there is any chance of that happening.
+    const resopAsync = prepareResource(
+        label, res, true, props,
+        (<CustomResourceOptions>opts).provider,
+        opts.parent && opts.parent.urn,
+        opts.dependsOn);
```

Comment:
note; this was very important, and fortunately tests caught this. We brought this up during discussion. But it was basically teh critical need to grab this data prior to async work happening. Otheriwse a resource could try depending ont he parent URN, but the parent URn had been updated to include waiting for the child resource. I switched to grabbing all the data eagerly and passin git along. I didn't even want to pass opts or opts.parent along in case someone might then read urn in some point in the future, causing a problem.